src/main/java/org/opensearch/knn/index/KNNIndexCache.java

```diff
@@ -312,38 +310,122 @@ public KNNIndexCacheEntry loadIndex(String indexPathUrl, String indexName) throw
         // the entry
         fileWatcher.init();
 
-        final KNNIndex knnIndex = KNNIndex.loadIndex(indexPathUrl, getQueryParams(indexName),
-                SpaceType.getSpace(KNNSettings.getSpaceType(indexName)).getValue());
+        // loadIndex from different library
+        String engineName;
+        Map<String, Object> parameters;
+        if (indexPathUrl.endsWith(KNNEngine.NMSLIB.getExtension())
+                || indexPathUrl.endsWith(KNNEngine.NMSLIB.getCompoundExtension())) {
+            engineName = KNNEngine.NMSLIB.getName();
+            parameters = ImmutableMap.of(
+                    KNNConstants.SPACE_TYPE, spaceType.getValue(),
+                    KNNConstants.HNSW_ALGO_EF_SEARCH, KNNSettings.getEfSearchParam(indexName)
+            );
+        } else {
+            throw new IllegalArgumentException("[KNN] Invalid engine type for path: " + indexPathUrl);
+        }
+
+        final long indexPointer = JNIService.loadIndex(indexPathUrl, parameters, engineName);
 
         // TODO verify that this is safe - ideally we'd explicitly ensure that the FileWatcher is only checked
         // after the guava cache has finished loading the key to avoid a race condition where the watcher
         // causes us to invalidate an entry before the key has been fully loaded.
         final WatcherHandle<FileWatcher> watcherHandle = resourceWatcherService.add(fileWatcher);
 
-        return new KNNIndexCacheEntry(knnIndex, indexPathUrl, indexName, watcherHandle);
+        return new KNNIndexCacheEntry(indexPointer, indexPathUrl, indexName, watcherHandle, engineName);
+    }
+
+    /**
+     * Execute a query on a given path. If the graph is not already loaded into memory, it will first load the graph
+     * into memory
+     *
+     * @param indexPathUrl path to graph file
+     * @param indexName name of OpenSearch index
+     * @param spaceType space type used for the graph
+     * @param query float vector used for query
+     * @param k number of neighbors to return
+     * @param engineName name of engine to use
+     * @return array of KNNQueryResult's
+     */
+    public KNNQueryResult[] queryIndex(String indexPathUrl, String indexName, SpaceType spaceType, float[] query,
+                                       int k, String engineName) {
+        KNNIndexCacheEntry knnIndexCacheEntry = getIndex(indexPathUrl, indexName, spaceType);
+        Lock readLock = knnIndexCacheEntry.getReadWriteLock().readLock();
+        readLock.lock();
+        try {
+            if (knnIndexCacheEntry.isClosed()) {
+                throw new IOException("Index is already closed");
```

Comment:
Minor: Add ```indexName, engineName``` to the message. If possible add spaceType as well. You may want to check other Exception places as well and see if we can add information to the error messages to help debug.